### PR TITLE
Reduce wallet page routing complexity

### PR DESCRIPTION
**Pull request by josheleonard**
Resolves https://github.com/brave/brave-browser/issues/22402
Resolves https://github.com/brave/brave-browser/issues/22588
Resolves https://github.com/brave/brave-browser/issues/22589

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by Douglashdaniel — CHANGES_REQUESTED** [review]
Pulled to test locally
Found some bugs that need to be addressed.

- Session Route is not being persisted. (example: While the wallet is locked, enter the route `brave://wallet/crypto/accounts`, it will redirect you to unlock the wallet. After you unlock the wallet it should direct you to `brave://wallet/crypto/accounts` not `brave://wallet/crypto/portfolio`

- `Add account` button is not working on the `Asset Detail` page.

- While the wallet is unlocked go to `brave://wallet/crypto/portfolio/braveisawesome`, if the token doesn't exist it should redirect back to `brave://wallet/crypto/portfolio/` which it looks like it's trying to do, but `/braveisawesome` stays persisted and the `Nav` bar disappears.

- `Restore` button isn't working on the `brave://wallet/crypto/unlock` page

- After creating a `password` for a new `wallet` it shows the `restore` page but the route should be `brave://wallet/crypto/backup-wallet`

- After creating a `wallet` the `skip` button isn't working on the `restore` page

- After creating a `wallet` the `recovery phrase` is empty on the next step.

**Comment by josheleonard** [comment]
> Pulled to test locally Found some bugs that need to be addressed.
> 
> * Session Route is not being persisted. (example: While the wallet is locked, enter the route `brave://wallet/crypto/accounts`, it will redirect you to unlock the wallet. After you unlock the wallet it should direct you to `brave://wallet/crypto/accounts` not `brave://wallet/crypto/portfolio`
> * `Add account` button is not working on the `Asset Detail` page.
> * While the wallet is unlocked go to `brave://wallet/crypto/portfolio/braveisawesome`, if the token doesn't exist it should redirect back to `brave://wallet/crypto/portfolio/` which it looks like it's trying to do, but `/braveisawesome` stays persisted and the `Nav` bar disappears.
> * `Restore` button isn't working on the `brave://wallet/crypto/unlock` page
> * `Restore` button isn't working on the `brave://wallet/crypto/onboarding` page
> * After creating a `password` for a new `wallet` it shows the `backup-wallet` page but the route should be `brave://wallet/crypto/backup-wallet`
> * After creating a `wallet` the `skip` button isn't working on the `backup-wallet` page
> * After creating a `wallet` the `recovery phrase` is empty on the next step.

Fixed in commit: c27cddd

**Review by Douglashdaniel — CHANGES_REQUESTED** [review]
Pulled your latest, few bugs found.
- Login to the wallet and paste `brave://wallet/crypto/backup-wallet` into the address bar, click continue, `Recover Phrase` is not returned.
-  Login to the wallet and paste brave://wallet/crypto/portfolio/ETH into the address bar, is redirected to `Portfolio` page. Should load the `ETH` asset details page.
- Go to the `Accounts` tab and click on your wallet, copy the route in the address bar example: `brave://wallet/crypto/accounts/0xaea1FFa9600c7812444Eee7ad4F3D17c2d7fAfBa`. Now go back to the portfolio page and paste that route back and hit enter. Blank page with console errors.
- Go to the `Accounts` tab and click `Create account`, create an account. The modal should dismiss after account is created.
- Go to the `Accounts` tab and click `Import account`, import an account. The modal should dismiss after account is imported.
- Go to the `Accounts` tab and click `Import account`, select accounts and click `Add checked accounts`. The modal should dismiss after accounts are added.

File: components/brave_wallet_ui/components/desktop/popup-modals/add-account-modal/add-hardware-account-modal.tsx [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+export const AddHardwareAccountModal = () => {
+  // routing
+  const history = useHistory()
+  const { accountTypeName } = useParams<Params>()
+
+  // redux
+  const dispatch = useDispatch()
+  const isFilecoinEnabled = useSelector(({ wallet }: { wallet: WalletState }) => wallet.isFilecoinEnabled)
+  const isSolanaEnabled = useSelector(({ wallet }: { wallet: WalletState }) => wallet.isSolanaEnabled)
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/components/desktop/popup-modals/add-account-modal/add-hardware-account-modal.tsx:45]: Can these be combined and deconstructed?

File: components/brave_wallet_ui/components/desktop/popup-modals/add-account-modal/add-imported-account-modal.tsx [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+  const passwordInputRef = React.useRef<HTMLInputElement>(null)
+
+  // routing
+  const history = useHistory()
+  // const { pathname: walletLocation } = useLocation()
+  const { accountTypeName } = useParams<Params>()
+
+  // redux
+  const isSolanaEnabled = useSelector(({ wallet }: { wallet: WalletState }) => wallet.isSolanaEnabled)
+  const isFilecoinEnabled = useSelector(({ wallet }: { wallet: WalletState }) => wallet.isFilecoinEnabled)
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/components/desktop/popup-modals/add-account-modal/add-imported-account-modal.tsx:64]: Same here?

File: components/brave_wallet_ui/components/desktop/views/portfolio/portfolio-asset.tsx [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+    dispatch(WalletPageActions.selectAsset({
+      asset: selectedAsset,
+      timeFrame: timeline
+    }))
+  }, [selectedAsset])
+
+  const goBack = React.useCallback(() => {
+    // if (nftMetadata) {
+    //   dispatch(WalletPageActions.updateNFTMetadata(undefined))
+    // }
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/components/desktop/views/portfolio/portfolio-asset.tsx:277]: ?

File: components/brave_wallet_ui/components/desktop/views/portfolio/portfolio-asset.tsx [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+      {!isNftAsset &&
+        <InfoColumn>
+          <AssetRow>
+            <AssetIconWithPlaceholder asset={selectedAsset} network={selectedAssetsNetwork} />
+            <AssetColumn>
+              <AssetNameText>{selectedAssetFromParams.name}</AssetNameText>
+              <NetworkDescription>{selectedAssetFromParams.symbol} on {selectedAssetsNetwork?.chainName ?? ''}</NetworkDescription>
+            </AssetColumn>
+          </AssetRow>
+          {/* <DetailText>{selectedAsset.name} {getLocale('braveWalletPrice')} ({selectedAsset.symbol})</DetailText> */}
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/components/desktop/views/portfolio/portfolio-asset.tsx:347]: Does this need to be removed?

File: components/brave_wallet_ui/components/desktop/wallet-onboarding/restore/index.tsx [hunk 5]
```diff
@@ -38,50 +43,73 @@ export const OnboardingRestore = () => {
 
   // redux
   const dispatch = useDispatch()
-  const {
-    isWalletCreated,
-    isWalletLocked
-  } = useSelector(({ wallet }: { wallet: WalletState }) => wallet)
-  const { invalidMnemonic } = useSelector(({ page }: { page: PageState }) => page)
+  const isWalletCreated = useSelector(({ wallet }: { wallet: WalletState }) => wallet.isWalletCreated)
+  const isWalletLocked = useSelector(({ wallet }: { wallet: WalletState }) => wallet.isWalletLocked)
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/components/desktop/wallet-onboarding/restore/index.tsx:47]: Can be combined?

File: components/brave_wallet_ui/components/desktop/popup-modals/add-account-modal/create-account-modal.tsx [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  const history = useHistory()
+  const { pathname: walletLocation } = useLocation()
+  const { accountTypeName } = useParams<Params>()
+
+  // redux
+  const dispatch = useDispatch()
+  const isFilecoinEnabled = useSelector(({ wallet }: { wallet: WalletState }) => wallet.isFilecoinEnabled)
+  const isSolanaEnabled = useSelector(({ wallet }: { wallet: WalletState }) => wallet.isSolanaEnabled)
+  const accounts = useSelector(({ wallet }: { wallet: WalletState }) => wallet.accounts)
+
```
[the remark this answers is not in the document]
Comment (josheleonard) [reply components/brave_wallet_ui/components/desktop/popup-modals/add-account-modal/create-account-modal.tsx:49]: selectors should only return one value to reduce chances or re-rendering

File: components/brave_wallet_ui/components/desktop/popup-modals/add-account-modal/add-imported-account-modal.tsx (hunk 2 above)
Comment (josheleonard) [reply components/brave_wallet_ui/components/desktop/popup-modals/add-account-modal/add-imported-account-modal.tsx:64]: selectors should only return one value to reduce chances or re-rendering

File: components/brave_wallet_ui/components/desktop/popup-modals/add-account-modal/add-hardware-account-modal.tsx (hunk 1 above)
Comment (josheleonard) [reply components/brave_wallet_ui/components/desktop/popup-modals/add-account-modal/add-hardware-account-modal.tsx:45]: selectors should only return one value to reduce chances or re-rendering

File: components/brave_wallet_ui/components/desktop/views/portfolio/portfolio-asset.tsx (hunk 3 above)
Comment (josheleonard) [reply components/brave_wallet_ui/components/desktop/views/portfolio/portfolio-asset.tsx]: left over from a reverted rebase. removed

File: components/brave_wallet_ui/components/desktop/views/portfolio/portfolio-asset.tsx (hunk 4 above)
Comment (josheleonard) [reply components/brave_wallet_ui/components/desktop/views/portfolio/portfolio-asset.tsx]: removed

File: components/brave_wallet_ui/common/async/handlers.ts [hunk 7]
```diff
@@ -232,6 +232,8 @@ handler.on(WalletActions.removeFavoriteApp.getType(), async (store: Store, appIt
   await refreshWalletInfo(store)
 })
 
+
```
Comment (muliswilliam) [inline components/brave_wallet_ui/common/async/handlers.ts:235]: Remove extra line

**Review by Douglashdaniel — APPROVED** [review]
LGTM!

File: components/brave_wallet_ui/common/async/handlers.ts (hunk 7 above)
Comment (josheleonard) [reply components/brave_wallet_ui/common/async/handlers.ts]: removed

**Pull request merged**